rust/src/lib/dns.rs

```diff
@@ -0,0 +1,551 @@
+                    }
+                }
+
+                if cur_conf.server != Some(canonicalized_srvs)
+                    && !(cur_conf.server.is_none() && srvs.is_empty())
+                {
+                    return Err(NmstateError::new(
+                        ErrorKind::VerificationError,
+                        format!(
+                            "Failed to apply DNS config: desire name servers \
+                            {:?}, got {:?}",
+                            srvs,
+                            cur_conf.server.as_ref()
+                        ),
+                    ));
+                }
+            }
+            if let Some(schs) = conf.search.as_ref() {
+                let cur_conf = current.config.as_ref().ok_or_else(|| {
+                    NmstateError::new(
+                        ErrorKind::VerificationError,
+                        format!(
+                            "Failed to apply DNS config: desire {:?} got {:?}",
+                            self, current
+                        ),
+                    )
+                })?;
+                if cur_conf.search != Some(schs.to_vec())
+                    && !(cur_conf.search.is_none() && schs.is_empty())
+                {
+                    return Err(NmstateError::new(
+                        ErrorKind::VerificationError,
+                        format!(
+                            "Failed to apply DNS config: desire searches \
+                            {:?}, got {:?}",
+                            schs,
+                            cur_conf.search.as_ref()
+                        ),
+                    ));
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub(crate) fn merge_current(&mut self, current: &Self) {
+        if let Some(conf) = self.config.as_mut() {
+            if conf.is_purge() {
+                self.config = Some(DnsClientSetting {
+                    server: Some(Vec::new()),
+                    search: Some(Vec::new()),
+                    priority: None,
+                });
+            } else if let Some(cur_conf) = current.config.as_ref() {
+                if conf.server.is_none() {
+                    conf.server =
+                        Some(cur_conf.server.clone().unwrap_or_default());
+                }
+                if conf.search.is_none() {
+                    conf.search =
+                        Some(cur_conf.search.clone().unwrap_or_default());
+                }
+            }
+        } else {
+            self.config = current.config.clone();
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct DnsClientSetting {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub server: Option<Vec<String>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub search: Option<Vec<String>>,
+    #[serde(skip)]
+    // Lower is better
+    pub(crate) priority: Option<i32>,
```

Comment:
What about something like `DnsClientState`? It seems strange to me to talk about Settings on `src/lib/dns.rs`..

Reply:
Changed to `DnsClientState`